packages/fpdart/lib/src/effect.dart

```diff
@@ -0,0 +1,561 @@
+import 'dart:async';
+
+import './extension/future_or_extension.dart';
+import './extension/iterable_extension.dart';
+import 'exit.dart';
+import 'unit.dart' as fpdart_unit;
+
+part 'either.dart';
+part 'option.dart';
+
+typedef EffectGen<E, L> = ({
+  FutureOr<A> Function<A>(IEffect<E, L, A>) async,
+  A Function<A>(IEffect<E, L, A>) sync,
+});
+
+final class _EffectThrow<L> implements Exception {
+  final Cause<L> cause;
+  const _EffectThrow(this.cause);
+
+  @override
+  String toString() {
+    return "Effect.gen error: $cause";
+  }
+}
+
+EffectGen<E, L> _effectGen<E, L>(E env) => (
+      async: <A>(effect) => Future.sync(
+            () => effect.asEffect._unsafeRun(env).then(
+                  (exit) => switch (exit) {
+                    Left(value: final cause) => throw _EffectThrow<L>(cause),
+                    Right(value: final value) => value,
+                  },
+                ),
+          ),
+      sync: <A>(effect) {
+        final run = effect.asEffect._unsafeRun(env);
+        if (run is Future) {
+          throw _EffectThrow<L>(
+            Die.current(
+              Exception("gen.sync cannot execute async Effect"),
+            ),
+          );
+        }
+
+        return switch (run) {
+          Left(value: final cause) => throw _EffectThrow<L>(cause),
+          Right(value: final value) => value,
+        };
+      },
+    );
+
+typedef DoFunctionEffect<E, L, A> = FutureOr<A> Function(
+  EffectGen<E, L> $,
+);
+
+abstract interface class IEffect<E, L, R> {
+  const IEffect();
+  Effect<E, L, R> get asEffect;
+}
+
+final class Effect<E, L, R> extends IEffect<E, L, R> {
+  final FutureOr<Exit<L, R>> Function(E env) _unsafeRun;
+  const Effect._(this._unsafeRun);
+
+  @override
+  Effect<E, L, R> get asEffect => this;
+
+  @override
+  String toString() {
+    return "Effect(${_unsafeRun.runtimeType})";
+  }
+
+  /// {@category execution}
+  R runSync(E env) {
+    try {
+      final result = _unsafeRun(env);
+      if (result is Future) {
+        throw Die.current(
+          Exception("runSync cannot execute async Effect"),
+        );
+      }
+
+      return switch (result) {
+        Left(value: final cause) => throw cause,
+        Right(value: final value) => value,
+      };
+    } on Cause<L> {
+      rethrow;
+    } catch (error, stackTrace) {
+      throw Die(error, stackTrace);
```

Comment:
I implemented `catchAllCause` (called `catchCause`) by making the error channel `Never`:
https://github.com/SandroMaglione/fpdart/blob/f02afe3d4eb4bde8ca45ee769def2b801c0abfa1/packages/fpdart/lib/src/effect.dart#L547-L558

When running then the idea is to `rethrow` `Cause` or throw `Die` otherwise.

What do you think about this approach?

Reply:
The problem is that you need to catch defects and convert them to Die for every Effect, not just when you call a run* method.

Reply:
I don’t understand this point. Do you mean that there should be a "catch" every time `_unsafeRun` is executed? Could you give me an example where the current approach doesn’t work?

Reply:
This should return "rescued!"

```dart
import "package:fpdart/fpdart.dart";

void main() {
  final result = Effect<Null, Never, String>.functionSucceed(() {
    throw "fail";
    return "hello!";
  }).catchCause((cause) => Effect.succeed("rescued!")).runSync();

  print(result);
}
```

Reply:
Got it, I [updated the code](https://github.com/SandroMaglione/fpdart/commit/f343b0314c53a45b8c050294b54fff82f64b9761) to create effects from a safe `Effect.from`